server_admin/topics/authentication/x509.adoc

```diff
@@ -209,6 +209,61 @@ image:images/x509-directgrant-flow.png[]
+
+If {project_name} is behind the reverse proxy, you usually need to configure alternative provider of the `x509cert-lookup` SPI in `{project_dirref}/standalone/configuration/standalone.xml` file. Besides the `default` provider, which lookups the certificate from the HTTP header, we have 2 builtin providers: `haproxy` and `apache`.
+
+===== HAProxy certificate lookup provider
+
+This provider can be used when your {project_name} server is behind the HAProxy reverse proxy. The provider needs to be configured like this:
+
+[source,xml]
+----
+<spi name="x509cert-lookup">
+    <default-provider>haproxy</default-provider>
+    <provider name="haproxy" enabled="true">
+        <properties>
+            <property name="sslClientCert" value="SSL_CLIENT_CERT"/>
+            <property name="sslCertChainPrefix" value="CERT_CHAIN"/>
+            <property name="certificateChainLength" value="10"/>
+        </properties>
+    </provider>
+</spi>
+----
+
+In this example configuration, the client certificate will be lookup from the HTTP header `SSL_CLIENT_CERT` and the other certificates from it's chain will be lookup from the HTTP headers like `CERT_CHAIN_0` , `CERT_CHAIN_1`, ..., `CERT_CHAIN_9` . The attribute `certificateChainLength` is the maximum length of the chain, so the last one tried attribute would be `CERT_CHAIN_9` .
+
+Consult the HAProxy documentation for the details how the HTTP Headers for the client certificate and client certificate chain can be configured and what are their proper names.
+
+===== Apache certificate lookup provider
+
+This provider can be used when your {project_name} server is behind the Apache reverse proxy. The provider needs to be configured like this:
+
+[source,xml]
+----
+<spi name="x509cert-lookup">
+    <default-provider>apache</default-provider>
+    <provider name="apache" enabled="true">
+        <properties>
+            <property name="sslClientCert" value="SSL_CLIENT_CERT"/>
+            <property name="sslCertChainPrefix" value="CERT_CHAIN"/>
+            <property name="certificateChainLength" value="10"/>
+        </properties>
+    </provider>
+</spi>
+----
+
+The configuration is same like for `haproxy` provider. Again, consult the Apache documentation for the details how the HTTP Headers for the client certificate and client certificate chain can be configured and what are their names.
+
+===== Other reverse proxy implementations
+
+We don't have builtin support for the other reverse proxy implementations. However it's possible that other reverse proxies can behave
+in similar way like `apache` or `haproxy` and some of those providers can be used. Alternatively you may need to create your own implementation of the `org.keycloak.services.x509.X509ClientCertificateLookupFactory` and `org.keycloak.services.x509.X509ClientCertificateLookup` provider. See the link:{developerguide_link}[{developerguide_name}] for the details how to add your own provider.
+
```

Comment:
We do not have built-in support for other reverse proxy implementations. However, it is possible that other reverse proxies can be made to behave in a similar way to `apache` or `haproxy` and that some of those providers can be used. If you want to try, you may need to create your own implementation of the `org.keycloak.services.x509.X509ClientCertificateLookupFactory` and `org.keycloak.services.x509.X509ClientCertificateLookup` provider. See the link:{developerguide_link}[{developerguide_name}] for the details on how to add your own provider.